python/ray/data/dataset_pipeline.py

```diff
@@ -642,6 +647,11 @@ def __iter__(self):
 
         return EpochDelimitedIterator(self)
 
+    def _peek(self) -> (Dataset[T], Iterator[Dataset[T]]):
+        self._optimize_stages()
+        dataset_iter = PipelineExecutor(self)
+        return next(dataset_iter), dataset_iter
```

Comment:
A slightly cleaner interface for `._peek()` would be if it returned the peeked at dataset and took care of the dataset/executor caching under-the-hood:
```suggestion
    def _peek(self) -> Dataset[T]:
        if self._first_dataset is not None:
            self._optimize_stages()
            self._dataset_iter = PipelineExecutor(self)
            self._first_dataset = next(self._dataset_iter)
        return self._first_dataset
```
This way the "peekers" don't have to worry about dataset/executor caching, and this also allows for multiple peeks at the same pipeline without launching another executor. Here's what `.schema()` could look like:
```python
    def schema(self, fetch_if_missing: bool = False):
        return self._peek().schema(fetch_if_missing=fetch_if_missing)
```

Reply:
Looking good, done!